File: pqc_cpp/number_theory/mod_exp.cpp
#include "mod_exp.hpp"

uint64_t modular_exponentiation(const uint64_t alpha, const uint64_t b, const uint64_t mu) {
    __uint128_t a = __uint128_t(alpha);
    __uint128_t m = __uint128_t(mu);

    a%=m;


    if (b==0) {
        return 1;
    }
    else {
        __uint128_t result = __uint128_t(1);
        for (uint64_t i = 0; i < 63 - std::countl_zero(b); ++i) {
            uint64_t bit = (b >> i) &1;

            if (bit == 1) {
                result = (result * a) % m;
            }
        a = (a*a)%m;
        }
    return uint64_t(result);
    }

}

File: pqc_rust/src/number_theory/mod_exp.rs
pub fn modular_exponentiation(a: u64, b: u64, m: u64) -> u64 {
    // Exponentiation by squaring
    let mut alpha: u128 = a as u128;
    let mu: u128 = m as u128;
    alpha%=mu;
    let mut result:u128 = 1;

    //the case division here makes it so that we can save one computation of updating a by squaring it
    if b==0 {
        result.try_into().unwrap()
    }
    else {
        for i in 0..63-b.leading_zeros() {
            let bit = (b >> i) & 1;
        
            if bit == 1 {
                result = (result * alpha) % mu;
            }
        
            alpha = ( alpha*alpha )% mu;
        
        }
        ((result*alpha)%mu).try_into().unwrap()
    }
    
    
}
